### DISCUSSION TITLE
VK_ERROR_INCOMPATIBLE_DRIVER

### DISCUSSION THREAD

**Discussion by bursica** [oobabooga/text-generation-webui#4221] [Q&A] [answered]
I get an error after Kernel update:
Error: ../src/intel/vulkan_hasvk/anv_device.c:671: Unable to open device /dev/dri/renderD128: Permission denied (VK_ERROR_INCOMPATIBLE_DRIVER)
Error: ../src/intel/vulkan/anv_device.c:851: Unable to open device /dev/dri/renderD128: Permission denied (VK_ERROR_INCOMPATIBLE_DRIVER)
2023-10-07 20:49:00 WARNING:skip module injection for FusedLlamaMLPForQuantizedModel not support integrate without triton yet.

Thank you for any ideas.

**Comment by bursica** [comment, marked as answer]
To answer with my findings:

It may be also connected with installed Vulkan mesa drivers in case of multiple GPUs, when one is integrated. Result, wrong GPU is used.

First option (can also be in ~/.bashrc):
export CUDA_VISIBLE_DEVICES=0        # when your compute GPU is 0

With only that change when running text-generation-webui it will throw info: "MESA-INTEL: warning: Haswell Vulkan support is incomplete".
If you dont' use mesa-vulkan then:
sudo apt-get purge mesa-vulkan-drivers

Hope this will help someone with similar issues.

**Comment by FrankVanris2** [comment]
Ok I had the same issue, and I was able to resolve it with these steps:

1. **Check Vulkan Driver Installation:** Ensure that the Vulkan drivers are properly installed and up-to-date:
```bash
sudo apt update
sudo apt install --reinstall libvulkan1 vulkan-tools
```

2. **Set Permissions Correctly:** Double-check the permissions on the `/dev/dri/renderD128` device:
```bash
sudo chmod 666 /dev/dri/renderD128
sudo chown root:render /dev/dri/renderD128
```

3. **Add User to Render Group:** Ensure your user is part of the `render` group:
```bash
sudo usermod -aG render $USER
```

4. **Restart the Service:** Restart the display manager or the service that uses Vulkan:
```bash
sudo systemctl restart display-manager
```

5. **Set Environment Variable:** Set the `LIBGL_ALWAYS_INDIRECT=1` environment variable to avoid direct rendering issues:
```bash
export LIBGL_ALWAYS_INDIRECT=1
```
6. **Purge and Reinstall Mesa Vulkan Drivers:** If the issue persists, try purging and reinstalling the Mesa Vulkan driers:
```bash
sudo apt purge mesa-vulkan-drivers
sudo apt install mesa-vulkan-drivers
```
Now after these steps whatever you were trying to do should work. I had the same issue because on my Raspberry pi with ROS2 I was trying to run this command: `ros2 run turtlesim turtlesim_node`

And when I was trying to do that I was getting these issues:
```bash
ubuntu@followbot-desktop:~$ ros2 run turtlesim turtlesim_node
[INFO] [1731691131.987738381] [turtlesim]: Starting turtlesim with node name /turtlesim
[INFO] [1731691132.013147640] [turtlesim]: Spawning turtle [turtle1] at x=[5.544445], y=[5.544445], theta=[0.000000]
TU: error: ../src/freedreno/vulkan/tu_knl.cc:208: failed to open device /dev/dri/renderD128 (VK_ERROR_INCOMPATIBLE_DRIVER)
Opening /dev/dri/renderD128 failed: Permission denied
MESA: error: ZINK: vkEnumeratePhysicalDevices failed (VK_ERROR_INITIALIZATION_FAILED)
MESA: error: ZINK: failed to choose pdev
glx: failed to create drisw screen
```

but after doing these steps I was finally able to get it to run without any conflicts. And as for the Mesa errors and stuff, the above steps resolves everything else because it just I guess started with renderD128 being denied permission.

Anyways hopes this helps for anyone. It took me to long to get this down. I was using ROS2 Jazzy on a Raspberry Pi 4B with Debian 24.04 LTS. So if anyone has this issue in the future here you go.

**Reply by FrankVanris2** [reply]
omg I forgot to mention that after you do all of these steps YOU MUST DO `sudo reboot` in order for your changes to make a difference.
